Teambuilder: Generate user samples from boxes

im-tofa opened the pull request:
This is an implementation of [a suggestion to display user sample sets from boxes](https://www.smogon.com/forums/threads/teambuilder-improvement-user-sets.3730629/).

The first time a user clicks "Import/Export" in a set view of the teambuilder, this addition will scan through the teambuilder for any boxes with the appropriate format, and cache all sets as user samples. These will then be displayed as buttons. User samples with duplicate names receive a number suffix. 

The cache for a given format is cleared if the user exits a box in said format. Caching reduces repetitive scans each time the user wants to import a set.

I did a quick "performance" test by creating 5000 teams, does not cause hiccups for me at least. Tested on Firefox. The bottleneck, if any, has been the smogon samples fetched using HTTP.

Let me know if I can improve the code quality or functionality, thanks!

This is what it looks like in my test client, see the bottom of the screenshot:
![](https://github.com/smogon/pokemon-showdown-client/assets/79044321/fa04c5b1-22f5-4432-bd04-76807df22837)

This was the corresponding box:
![](https://github.com/smogon/pokemon-showdown-client/assets/79044321/4cbf209c-e47b-4441-b304-dd774724546f)

im-tofa commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -1789,34 +1792,71 @@
 				self.importSetButtons();
 			}, 'text');
 		},
+		getUserSets: function (format) {
```
Should perhaps place an upper limit on the number of sets per Pokémon? Just so someone doesn't have like 45 samples of a Pokémon in a given format, idk what would be a reasonable limit though

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -1792,31 +1797,76 @@
+					for(const pokemon of setList) {
+						var name = pokemon.name + " " + (duplicateNameIndices[pokemon.name] || "");
+						var sets = this.userSets[format][pokemon.species];
+						this.userSets[format][pokemon.species] = $.extend(sets || {}, {[name] : pokemon});
```
`{[name] : pokemon}` is not ES3, you'll need to do this manually.

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -1792,31 +1797,76 @@
+					const setList = Storage.unpackTeam(team.team);
+					for(const pokemon of setList) {
+						var name = pokemon.name + " " + (duplicateNameIndices[pokemon.name] || "");
+						var sets = this.userSets[format][pokemon.species];
+						this.userSets[format][pokemon.species] = $.extend(sets || {}, {[name] : pokemon});
```
```suggestion
						var sets = this.userSets[format][pokemon.species] || {};
						sets[name] = pokemon;
						this.userSets[format][pokemon.species] = sets;
```

I think this is what you're trying to do?

im-tofa replied:
Exactly, thanks! :)

Zarel commented on play.pokemonshowdown.com/js/client-teambuilder.js:
```diff
@@ -1792,31 +1797,76 @@
+		},
+		clearCachedUserSetsIfNecessary: function (format) {
+			// clear cached user sets if we have just been in a box for given format
+			if(this.curTeam?.capacity === 24 && this.userSets?.[format]) {
```
`?.` is also not ES3.

Is something wrong with our CI? It should be catching this...

Zarel replied:
Oh, CI needs to be manually approved.

im-tofa commented:
Thanks for the feedback @Zarel! I've fixed the linting issues now (no more use of `?.`, `for-of`, `let`, `const` or `{ [name] : pokemon }`), made sure to try it out locally first now which I should've done the first time, my bad :)

After some thinking, I realised this feature is not very easy to discover or "self-explanatory". In an attempt to improve this, I changed the displayed text from "User sets" to "Box sets" and added a hint if no box sets are available for the format, see bottom of below screenshot. Let me know if that's not desired or if there are better alternatives:

![bild](https://github.com/smogon/pokemon-showdown-client/assets/79044321/7446c05e-7830-4a6b-871e-73a6bf693e34)
